addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js

```diff
@@ -360,6 +369,108 @@ class EnvironmentScConnectionService extends Service {
     return { password, networkInterfaces: this.toNetworkInterfaces(instanceInfo) };
   }
 
+  async createPrivateSageMakerUrl(requestContext, envId, connection, presign_retries = 10) {
+    const [lockService] = await this.service(['lockService']);
+    const signedURL = await lockService.tryWriteLockAndRun({ id: `${envId}presign` }, async () => {
+      if (!(_.toLower(_.get(connection, 'type', '')) === 'sagemaker')) {
+        throw this.boom.badRequest(
+          `Cannot generate presigned URL for non-sagemaker connection ${connection.type}`,
+          true,
+        );
+      }
+      const [environmentScService] = await this.service(['environmentScService']);
+      const iam = await environmentScService.getClientSdkWithEnvMgmtRole(
+        requestContext,
+        { id: envId },
+        { clientName: 'IAM', options: { apiVersion: '2017-07-24' } },
+      );
+      const currentPolicyResponse = await iam
+        .getRolePolicy({
+          RoleName: connection.role,
+          PolicyName: connection.policy,
+        })
+        .promise();
+
+      // Construct new statement which will allow the caller IP address permission to generate the presigned URL
+      const currentIpAddress = await fetch('http://checkip.amazonaws.com/').then(function(res) {
+        return res.text();
+      });
+      const newStatement = {
+        Effect: 'Allow',
+        Action: 'sagemaker:CreatePresignedNotebookInstanceUrl',
+        Resource: `${connection.notebookArn}`,
+        Condition: {
+          IpAddress: {
+            'aws:SourceIp': `${currentIpAddress.trim()}/32`,
+          },
+        },
+      };
+      const policyToUpdate = JSON.parse(decodeURIComponent(currentPolicyResponse.PolicyDocument));
+      let policyToUpdateStatement = policyToUpdate.Statement;
+      if (_.isArray(policyToUpdateStatement)) {
+        policyToUpdateStatement.push(newStatement);
+      } else {
+        policyToUpdateStatement = [policyToUpdateStatement, newStatement];
+        policyToUpdate.Statement = policyToUpdateStatement;
+      }
+      const putRolePolicyParams = {
+        RoleName: connection.role,
+        PolicyName: connection.policy,
+        PolicyDocument: JSON.stringify(policyToUpdate),
+      };
+      await iam.putRolePolicy(putRolePolicyParams).promise();
+
+      // wait for permission to propagate before using the role
+      await sleep(1000);
```

Comment:
We might want to do a while loop to deterministically wait for the connection to be available. Maybe max wait time of 10 seconds while we keep trying to get the url. If the request fails with a permission error, wait one second and try again. 

We should probably also log if we fail to generate the url after 10 seconds.

Reply:
Actually this sleep is completely unnecessary. The call to generate the URL is happening with retries (exponential backoff). So I will just remove it.